plutus-tx-plugin/src/PlutusTx/Compiler/Expr.hs

```diff
@@ -848,6 +857,280 @@ compileExpr e = withContextM 2 (sdToTxt $ "Compiling expr:" GHC.<+> GHC.ppr e) $
+                _ -> const Nothing
+            ),
+            ( ''GHC.Classes.Ord
+            , \case
+                -- `Ord Integer`
+                [ty] | ty `GHC.eqType` GHC.integerTy -> \(cls, v, args) ->
+                    Just $ compileDict cls v args Dictionary.Integer.selectorsOrd
+                -- `Ord [a]`
+                [GHC.splitTyConApp_maybe -> Just (tc, [_targ])] | tc == GHC.listTyCon ->
+                    \(cls, v, args) ->
+                        Just $ compileDict cls v args Dictionary.List.selectorsOrd
+                _ -> const Nothing
+            )
+          ]
+
+-- | Compile a class dictionary using alternative method definitions.
+--
+-- This is useful because some class instances' methods either don't have unfoldings,
+-- or cannot be directly compiled. Or we may want to use an alternative definition
+-- because it is more efficient.
+--
+-- For instance, the unfolding of `== @Integer` contains primitive operations that the
+-- plugin cannot handle, so we compile the dictionary of `Eq Integer` by using an
+-- alternative `==` from `PlutusTx.Compiler.Dictionary.Integer`.
+--
+-- Another example is `== [a]` which does not have unfolding. So we compile the dictionary
+-- of `Eq [a]` by using an alternative `==` from `PlutusTx.Compiler.Dictionary.List`.
+compileDict ::
+    CompilingDefault uni fun m ann =>
+    -- | The class whose dictionary we are compiling
+    GHC.Class ->
+    -- | Dictionary data constructor
+    GHC.Var ->
+    -- | Arguments to the dictionary data constructor
+    [GHC.CoreExpr] ->
+    -- | A list of dictionary method names where alternative definitions exist
+    [TH.Name] ->
+    m (PIRTerm uni fun)
+compileDict cls v args0 txSelNames = do
+    let selNames = GHC.getName <$> GHC.classAllSelIds cls
+        (otherArgs, selArgs0) = splitAtEnd (length selNames) args0
+    selArgs <- for (zip selNames selArgs0) $ \(n, arg) -> do
+        let (_argFun, argArgs) = GHC.collectArgs arg
+            txName =
+                find
+                    ( \n' ->
+                        GHC.occNameString (GHC.nameOccName n)
+                            == TH.nameBase n'
+                    )
+                    txSelNames
+        maybe
+            (pure arg)
+            ( pure . flip GHC.mkCoreApps argArgs . GHC.Var
+                <=< lookupId
+                <=< thNameToGhcNameOrFail
+            )
+            txName
+    case NE.nonEmpty (otherArgs <> selArgs) of
+        Nothing -> throwPlain $ UnsupportedError "Empty class"
+        Just args ->
+            -- Here we cannot apply `compileExpr` to
+            -- `GHC.mkCoreApps (GHC.Var v) (otherArgs <> selArgs)` directly, because
+            -- it will loop. We therefore separately compile the last argument.
+            PIR.Apply annMayInline
+                <$> compileExpr (GHC.mkCoreApps (GHC.Var v) (NE.init args))
+                <*> compileExpr (NE.last args)
+
+splitGhcName :: GHC.Name -> (Maybe String, String)
```

Comment:
I'm just uncovinced you need this at all if you use a function and the appropriate `Eq` instances.